### PR TITLE
Validate ZIP code for facility state

### LINKED ISSUE #3044: Validate that zip code is within selected state for facilities
### Description
There's currently no validation on our facility form that the user-input state matches their zip code. We do have SmartyStreets try to tell the user if their address isn't found in the system, but it doesn't seem to be offering suggestions when the state and zip don't match. This is leading to issues in reporting tests to states (seems to have happened a few times in VT, when users selected WA by mistake - we've gotten reports from both VT and WA about this.)

### Proposed solution 
Facility address validation should be updated to **block submission** of facilities where the given zip code doesn't match the state input. For v0, this can just be an error thrown back to the frontend with a note that the address needs to be fixed. For future iterations, it may be nice to have users input their ZIP first and then autofill the state field accordingly.

*** Also note that for now, this should not affect patient address validation as we can be more lenient there and don't want to block users' workflows.

### Additional context
https://skylight-hq.slack.com/archives/C019PRYMR0Q/p1638202077149900

### PR DESCRIPTION
**Pull request by nathancrtr**
## Related Issue or Background Info

- Closes #3044

## Changes Proposed
- Prevent address validation if the provided ZIP code is not in the selected state
    - For facilities _only_ at this time
    - Not worrying about patient/provider address validation for now

## Additional Information
- The structure of the existing `FacilityForm.test.tsx` was not friendly to the mocking I wanted to do here
    - Broke up the module-level mocking of `smartyStreets` API
    - Added more `describe` blocks for clarity and better control of state

## Screenshots / Demos
https://user-images.githubusercontent.com/27730981/147703062-a12a4394-13c9-4753-b95b-d7bfc83ec431.mov
* **Note**: in the above clip, `90000` is a completely invalid ZIP code (i.e. it does not exist anywhere in the US) and `12345` is invalid _for the selected state_

## Checklist for Author and Reviewer
- [ ] @Rebecca-LM if there's an error message you like better than `Invalid ZIP code for the selected state` drop me a comment here and I'll update it 👍 

### UI
- [ ] Any changes to the UI/UX are approved by design 
- [ ] Any new or updated content (e.g. error messages) are approved by design 

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [x] Database changes are submitted as a separate PR
  - [x] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [x] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [x] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
  - [x] Each new changeset has a corresponding [tag](https://docs.liquibase.com/change-types/community/tag-database.html)
- [x] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] DevOps team has been notified if PR requires ops support
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

**Review by emmastephenson — APPROVED** [review]
Thanks Nathan, looks great! Appreciate all the test coverage 👍

File: frontend/src/app/utils/smartyStreets.ts [hunk 1]
```diff
@@ -1,23 +1,13 @@
...
-const getClient = () => {
-  if (process.env.REACT_APP_SMARTY_STREETS_KEY === undefined) {
-    throw new SmartyStreetsError("Missing REACT_APP_SMARTY_STREETS_KEY");
-  }
-  const credentials = new core.SharedCredentials(
-    process.env.REACT_APP_SMARTY_STREETS_KEY
-  );
-  return core.buildClient.usStreet(credentials);
-};
+// cf. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/11436c5a19fc6aabfd6b8d93b37dac38b4ab9bc2/types/smartystreets-javascript-sdk/index.d.ts#L625
```
Comment (zdeveloper) [inline frontend/src/app/utils/smartyStreets.ts:6]: what does the comment do?

**Review by zdeveloper — APPROVED** [review]
LGTM! with a question

**Review by JSkeets — APPROVED** [review]
LGTM, lots of work invested into testing it appears, it's a fine art.

File: frontend/src/app/utils/smartyStreets.test.ts [hunk 2]
```diff
@@ -1,72 +1,188 @@
...
+    it("returns `false` if ZIP code is not valid for state", () => {
+      const sut = isValidZipCodeForState("CA", validZipCodeResult);
+
+      expect(sut).toBe(false);
+    });
+
+    it("returns `true` if ZIP code is valid for state", () => {
+      const sut = isValidZipCodeForState("NY", validZipCodeResult);
+
+      expect(sut).toBe(true);
```
Comment (nickclyde) [inline frontend/src/app/utils/smartyStreets.test.ts:185]: Great test coverage!!

**Review by nickclyde — APPROVED** [review]

File: frontend/src/app/utils/smartyStreets.ts (hunk 1 above)
Comment (nathancrtr) [reply frontend/src/app/utils/smartyStreets.ts:6]: Ah, looks like I deleted my helpful comment explaining it. Basically, the provided `usZipcode.Result` type stubbornly insists that `status` and `reason` should be present, but in fact those properties only appear on responses for invalid ZIP codes.

**Comment by Rebecca-LM** [comment]
@nathancrtr I'm just getting to tickets from over the holidays! In response to your question above: if there's an error message you like better than "Invalid ZIP code for the selected state" drop me a comment here and I'll update it 

I'd please change to "Invalid ZIP code for this state"

No rush but lmk if you can update!

**Pull request merged**